core/src/main/java/org/sonarsource/sonarlint/core/util/ProgressWrapper.java

```diff
@@ -19,22 +19,34 @@
  */
 package org.sonarsource.sonarlint.core.util;
 
+import javax.annotation.Nullable;
 import org.sonarsource.sonarlint.core.client.api.common.ProgressMonitor;
 import org.sonarsource.sonarlint.core.client.api.exceptions.CanceledException;
 
-import javax.annotation.Nullable;
-
 public class ProgressWrapper {
-  private ProgressMonitor handler;
 
-  public ProgressWrapper(@Nullable ProgressMonitor handler) {
+  private final ProgressMonitor handler;
+  private final float offset;
+  private final float factor;
+
+  private ProgressWrapper(float offset, float factor, @Nullable ProgressMonitor handler) {
+    this.offset = offset;
+    this.factor = factor;
     if (handler == null) {
       this.handler = new NoOpProgressMonitor();
     } else {
       this.handler = handler;
     }
   }
 
+  public ProgressWrapper(@Nullable ProgressMonitor handler) {
+    this(0.0f, 1.0f, handler);
+  }
+
+  public ProgressWrapper subProgress(float fromFraction, float toFraction) {
+    return new ProgressWrapper(offset + fromFraction * factor, (toFraction - fromFraction) * factor, handler);
```

Comment:
@henryju woudln't it be better for the subProgress to have a part of the message that is kept? The code to where the subProgress is passed might not know the context in which it is executed. Example, `subProgress("Downloading QPs", 0.4f, 0.6f); ` and then when calling `subProgress.setProgress("page 1/10", 0.1f)`, we would have the message "Download QPs (page 1/10)".